pyro/infer/mcmc/hmc.py

```diff
@@ -201,16 +204,20 @@ def _find_reasonable_step_size(self, z):
         return step_size
 
     def _sample_r(self, name):
-        r_dist = self._adapter.r_dist
-        r_flat = pyro.sample(name, r_dist)
         r = {}
-        pos = 0
-        for name, param in sorted(self.initial_params.items()):
-            next_pos = pos + param.numel()
-            r[name] = r_flat[pos:next_pos].reshape(param.shape)
-            pos = next_pos
-        assert pos == r_flat.size(0)
-        return r, r_flat
+        r_flat_dict = {}
+        for site_names in self.inverse_mass_matrix:
+            r_flat = pyro.sample(
+                name + "_" + ",".join(site_names), self._adapter.r_dist[site_names])
```

Comment:
This could result in name conflict, e.g. if we have names `loc_scale` and `loc` and `scale` (yes I've done that 😊). Do you think it's worth working around with positional naming, e.g.
```py
for i, site_name in enumerate(self.invere_mass_matrix):
    ... pyro.sample("{}_r_{}".format(name, i), ...)
```

Reply:
Good point! I was worried that there is a name conflict here. I think we can just simply use the name `"{}_{}".format(name, str(site_names))` here, so if `site_names` is `("x", "y", "z")`, the site will have name `name_('x', 'y', 'z')`.